codalab/rest/bundles.py

```diff
@@ -451,16 +458,67 @@ def _fetch_bundle_locations(bundle_uuid: str):
 )
 def _add_bundle_location(bundle_uuid: str):
     """
-    Adds a new BundleLocation to a bundle. Request body must contain the fields in BundleLocationSchema.
+    Adds a new BundleLocation to a bundle. If need to generate sas token, generate Azure SAS token and connection string. Request body must contain the fields in BundleLocationSchema.
 
     Query parameters:
-    - `bundle_uuid`: Bundle UUID corresponding to the new location
+    - `need_sas`: (Optional) Bool. If true, if will return SAS token
+    - `is_dir`: (Optional) Bool. Whether the uploaded file is directory.
     """
+    check_bundles_have_all_permission(local.model, request.user, [bundle_uuid])
+    need_sas = query_get_bool('need_sas', default=False)
+    is_dir = query_get_bool('is_dir', default=False)
+
+    bundle = local.model.get_bundle(bundle_uuid)
     new_location = BundleLocationSchema(many=True).load(request.json).data[0]
-    new_location["uuid"] = local.model.add_bundle_location(
-        new_location['bundle_uuid'], new_location['bundle_store_uuid']
+    logging.info(
+        f"Try to bypass server upload, need_sas: {need_sas}, is_dir: {is_dir}, new_location: {new_location}"
     )
-    return BundleLocationSchema(many=True).dump([new_location]).data
+    # Scenario 1: User does not sepcify destination store, but rest-server support using Azure as default storage.
+    # Should bypass server and upload to default Azure store.
+    if (
+        new_location.get('bundle_store_uuid', None) is None
+        and os.environ.get('CODALAB_ALWAYS_USE_AZURE_BLOB_BETA') == '1'
+    ):
+        # The rest-server use Azure as default storage. Use default azure path
+        local.model.update_bundle(
```

Comment:
@wwwjn I think this is where you need to change the PR. We should no longer check for CODALAB_ALWAYS_USE_AZURE_BLOB_BETA. Rather, we should check for CODALAB_DEFAULT_BUNDLE_STORE_NAME, and, in that case, we should create a bundle store.

Here's a simple (but just one) way to do it:
```
bundle_store_uuid = new_location.get('bundle_store_uuid', None)
if bundle_store_uuid is None and os.environ.get('CODALAB_DEFAULT_BUNDLE_STORE_NAME'):
    bundle_store_uuid = # get bundle store uuid from os.environ.get('CODALAB_DEFAULT_BUNDLE_STORE_NAME')
...
```